src/cached_object_store.rs

```diff
@@ -762,6 +817,260 @@ impl LocalCacheEntry for FsCacheEntry {
+    }
+}
+
+/// FsCacheEvictorInner manages the cache entries in an in-memory trie, and evict the cache entries
+/// when the cache size exceeds the limit. it uses a pick-of-2 strategy to approximate LRU, and evict
+/// the older file when the cache size exceeds the limit.
+///
+/// On start up, FsCacheEvictorInner will scan the cache folder to load the cache files into the in-memory
+/// trie cache_entries. This loading process is interleaved with the maybe_evict is being called, so the
+/// cache entries should be wrapped with Mutex<_>.
+#[derive(Debug)]
+struct FsCacheEvictorInner {
+    root_folder: std::path::PathBuf,
+    batch_factor: usize,
+    max_cache_size_bytes: usize,
+    cache_entries: Mutex<Trie<std::path::PathBuf, (SystemTime, usize)>>,
+    cache_size_bytes: AtomicU64,
+    db_stats: Arc<DbStats>,
+}
+
+impl FsCacheEvictorInner {
+    pub fn new(
+        root_folder: std::path::PathBuf,
+        max_cache_size_bytes: usize,
+        db_stats: Arc<DbStats>,
+    ) -> Self {
+        Self {
+            root_folder,
+            batch_factor: 10,
+            max_cache_size_bytes,
+            cache_entries: Mutex::new(Trie::new()),
+            cache_size_bytes: AtomicU64::new(0_u64),
+            db_stats,
+        }
+    }
+
+    // scan the cache folder, and load the cache entries into the in memory trie cache_entries.
+    // this function is only called on start up, and it's expected to runned interleavely with
+    // maybe_evict is being called.
+    pub async fn scan_entries(self: Arc<Self>) {
+        // walk the cache folder, record the files and their last access time into the cache_entries
+        let iter = WalkDir::new(&self.root_folder).into_iter();
+        for entry in iter {
+            let entry = match entry {
+                Ok(entry) => entry,
+                Err(err) => {
+                    warn!("evictor: failed to walk the cache folder: {}", err);
+                    continue;
+                }
+            };
+            if entry.file_type().is_dir() {
+                continue;
+            }
+
+            let metadata = match tokio::fs::metadata(entry.path()).await {
+                Ok(metadata) => metadata,
+                Err(err) => {
+                    warn!(
+                        "evictor: failed to get the metadata of the cache file: {}",
+                        err
+                    );
+                    continue;
+                }
+            };
+            let atime = metadata.accessed().unwrap_or(SystemTime::UNIX_EPOCH);
+            let path = entry.path().to_path_buf();
+            let bytes = metadata.len();
+
+            self.cache_entries
+                .lock()
+                .await
+                .insert(path, (atime, bytes as usize));
+            self.cache_size_bytes.fetch_add(bytes, Ordering::SeqCst);
+        }
+    }
+
+    pub async fn maybe_evict(&self, path: std::path::PathBuf, bytes: usize) -> usize {
```

Comment:
Since we're now storing the access time in memory, does it still make sense to do 2-choice eviction vs. an LRU of the access times? I guess we lose the memory efficiency of the the trie for path keys, but that seems fairly minimal. Plus, we could store only the filename rather than the complete path to reduce memory usage.

Reply:
this approach resembles JuiceFS's method:

1. when atime is enabled on the filesystem, it can restore access times after a process restart.
2. when the filesystem is mounted with noatime, it mostly utilizes in-memory access times for LRU, and regard the `mtime` as default.

the 2-choice eviction can be viewed as an LRU implementation based on access times. instead of using a linked list to track LRU, it uses randomization to approximate it. this is also similar to Redis's LRU implementation (https://redis.io/docs/latest/develop/reference/eviction/#approximated-lru-algorithm)

Reply:
> I guess we lose the memory efficiency of the the trie for path keys, but that seems fairly minimal

in my previous understanding, if we have two keys like:

```
/var/folders/tz/9f04ptmx4892t1p2bfjbvkdw0000gn/T/objstore_cache_test_evictor_9Rj5zY/file0
/var/folders/tz/9f04ptmx4892t1p2bfjbvkdw0000gn/T/objstore_cache_test_evictor_9Rj5zY/file1
```

their common prefix `"/var/folders/tz/9f04ptmx4892t1p2bfjbvkdw0000gn/T/objstore_cache_test_evictor_9Rj5zY/file"` will be reused in a trie and will not store the prefix duplicated in the memory, am i understanding this correctly? 😲

Reply:
Yes, that's how I understand it as well. SGTM on keeping 2-choice for now.